### PR TITLE
Refactor simulation architecture

### PR DESCRIPTION
**Pull request by joshwlambert**
This PR addresses #42 by replacing `.sim_bp_linelist()`, `.sim_clinical_linelist()` and `.sim_contacts_tbl()` with `.sim_internal()`.  `sim_linelist()`, `sim_contacts()`, and `sim_outbreak()` now all call `.sim_internal()`.

The `sim_utils.R` file has been renamed to `utils.R` as the internal simulation functions have been removed.

The function documentation has been updated, the design principles vignette now refers to `.sim_internal()` rather than `.sim_bp_linelist()`, and snapshots for the regression tests have been updated. 

A diagram of the {simulist} package architecture and call stack from exported functions is added to the design principles vignette.

### CODE REVIEW

**Comment by joshwlambert** [comment]
Will merge tomorrow, as these changes are needed to produce a reprex for an issue in another package.

File: tests/testthat/_snaps/sim_outbreak.md
Comment (Bisaloo) [inline tests/testthat/_snaps/sim_outbreak.md:1]: This is more than just the rowname change. Could you document in the `NEWS.md` that runs with the same seed will give different result compared to the latest release?

**Review by Bisaloo — COMMENTED** [review]
Interface is not changed so I've reviewed this summarily. 

Do we need to question the fact that snapshot tests have changed? Even if you confirm this change is expected and valid, please document it in the changelog.

File: R/sim_contacts.R [hunk 1]
```diff
@@ -65,24 +65,26 @@ sim_contacts <- function(contact_distribution,
...
+  contacts <- .sim_internal(
+    sim_type = "contacts",
     contact_distribution = contact_distribution,
     contact_interval = contact_interval,
     prob_infect = prob_infect,
+    onset_to_hosp = NULL,
+    onset_to_death = NULL,
+    hosp_risk = NULL,
+    hosp_death_risk = NULL,
+    non_hosp_death_risk = NULL,
```
[the remark this answers is not in the document]
Comment (joshwlambert) [reply R/sim_contacts.R]: Resolved in 1ed4e321267f58e77cf94cda4fa4fcab1ecc3772.

File: tests/testthat/_snaps/sim_outbreak.md
Comment (joshwlambert) [reply tests/testthat/_snaps/sim_outbreak.md:1]: These are not actually changing the data in the `<data.frame>`s. The `row.names` are changing and this is causing the git diffs to pick up differences that are not actually there. I've been through the snapshots and other than the `row.names` which are meant to be changed in this PR, the data is identical.

**Comment by joshwlambert** [comment]
During the review of this PR a diagram of the package architecture has been added to the design principles vignette. I've edited the original comment of this PR to include these changes.

**Pull request merged**